relay-cogs/src/cogs.rs

```diff
@@ -0,0 +1,336 @@
+    ///
+    /// fn process(cogs: &Cogs, item: &Item) {
+    ///     let mut token = cogs.timed(ResourceId::Relay, AppFeature::Unattributed);
+    ///
+    ///     // App feature is only known after some computations.
+    ///     if do_something(item) {
+    ///         token.update(AppFeature::Spans);
+    ///     } else {
+    ///         token.update(AppFeature::Transactions);
+    ///     }
+    /// }
+    /// ```
+    pub fn update<T: Into<AppFeatures>>(&mut self, features: T) {
+        self.features = features.into();
+    }
+}
+
+impl Drop for CogsToken {
+    fn drop(&mut self) {
+        let elapsed = self.start.elapsed();
+
+        for (feature, ratio) in self.features.weights() {
+            let time = elapsed.div_f32(ratio);
+            self.recorder.record(CogsMeasurement {
+                resource: self.resource,
+                feature,
+                value: Value::Time(time),
+            });
+        }
+    }
+}
+
+impl fmt::Debug for CogsToken {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("CogsToken")
+            .field("resource", &self.resource)
+            .field("features", &self.features)
+            .finish()
+    }
+}
+
+/// A collection of weighted [app features](AppFeature).
+///
+/// Used to attribute a single COGS measurement to multiple features.
+#[derive(Clone, Copy)]
+pub struct AppFeatures(EnumMap<16, AppFeature, NonZeroUsize>);
+
+impl AppFeatures {
+    /// Attributes all measurements to a single [`AppFeature`].
+    pub fn new(feature: AppFeature) -> Self {
+        Self::builder().weight(feature, 1).build()
+    }
+
+    /// Attributes the measurement to nothing.
+    pub fn none() -> Self {
+        Self::builder().build()
+    }
+
+    /// Returns an [`AppFeatures`] builder.
+    pub fn builder() -> AppFeaturesBuilder {
+        AppFeaturesBuilder(Self(Default::default()))
+    }
+
+    /// Merges two instances of [`AppFeatures`] and sums the contained weights.
+    pub fn merge(mut self, other: Self) -> Self {
+        for (feature, weight) in other.0.into_iter() {
+            if let Some(w) = self.0.get_mut(feature) {
+                *w = w.saturating_add(weight.get());
+            } else {
+                self.0.insert(feature, weight);
+            }
+        }
+
+        self
+    }
+
+    /// Returns an iterator yielding an app feature and it's associated weight
+    /// normalized to the total stored weights in the range between `0.0` and `1.0`.
+    ///
+    /// Used to divide a measurement by the stored weights.
+    pub fn weights(&self) -> impl Iterator<Item = (AppFeature, f32)> {
+        let total_weight: usize = self.0.into_iter().map(|(_, weight)| weight.get()).sum();
+
+        self.0.into_iter().filter_map(move |(feature, weight)| {
+            if total_weight == 0 {
+                return None;
+            }
+
+            let ratio = (weight.get() as f32 / total_weight as f32).clamp(0.0, 1.0);
```

Comment:
Thinking out loud:

Are all ratios equivalent? Example: a ratio of 0.5 represents a usage that may (not) be the same as on high usage. Does this also depend on the configuration of k8s or the cloud provider?

Reply:
The cogs stuff only works on measured, actual cost. If I rate limit metrics twice and the first time it may take 1ms and the second time 5 ms, each time that cost is divided according to the weights. For the first measurement 0.5ms will be attributed (0.5 ratio) and for the second one 2.5ms, which totals to 3ms.

At the end of the day you can sum up all measurements and calculate the final ratios of everything and these ratios you can then apply to your actual costs.